### Problem 4

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import scipy.stats as stats
import statsmodels.api as sm
import statsmodels.formula.api as smf
from statsmodels.formula.api import ols
import math
import statsmodels.api as sm
import statsmodels.formula.api as smf
from statsmodels.formula.api import ols

In [53]:
df = pd.read_csv('problem4.txt',sep= '\s+',header=None) # read the data set
df.columns = (['Y','X1','X2','X3'])
df.head() 

Y  X1  X2   X3
0  48  50  51  2.3
1  57  36  46  2.3
2  66  40  48  2.2
3  70  41  44  1.8
4  89  28  43  1.8

### a.

In [6]:

X1 = df['X1']
X2 = df['X2']
X3 = df['X3']
Y = df['Y']
n=len(X1)

In [11]:
matY_1 = np.array(Y)
print('Y= ',matY_1,'\n')

Y=  [48 57 66 70 89 36 46 54 26 77 89 67 47 51 57 66 79 88 60 49 77 52 60 86
 43 34 63 72 57 55 59 83 76 47 36 80 82 64 37 42 66 83 37 68 59 92] 



In [12]:
col1 = np.array(X1)
col2 = np.array(X2)
col3 = np.array(X3)
matJ = np.ones((len(X1)))
matX = np.stack( (matJ,col1,col2,col3),axis = 1)

In [13]:
X_X1 = np.dot(matX.T,matX)
print("X'X= ",X_X1,'\n')

X'X=  [[4.60000e+01 1.76600e+03 2.32000e+03 1.05200e+02]
 [1.76600e+03 7.13780e+04 9.00510e+04 4.10720e+03]
 [2.32000e+03 9.00510e+04 1.17846e+05 5.34470e+03]
 [1.05200e+02 4.10720e+03 5.34470e+03 2.44620e+02]] 



In [14]:
X_Y1 = np.dot(matX.T,matY_1)
print("X'Y= ",X_Y1, '\n')

X'Y=  [  2832. 103282. 140814.   6327.] 



In [15]:
Invert = np.linalg.pinv(X_X1) 

In [16]:
b1=Invert.dot(X_Y1)

In [17]:
b_trans_1 = np.transpose(b1)
b_trans_1

array([158.49125167,  -1.14161185,  -0.44200426, -13.47016319])

In [18]:
# double check the results by OLS 
from sklearn import linear_model
y=matY_1
x=matX
regr=linear_model.LinearRegression()
regr.fit(x,y)
print('Intercept:\n',regr.intercept_)
print('Coeficients:\n',regr.coef_ )

Intercept:
 158.49125166661733
Coeficients:
 [  0.          -1.14161185  -0.44200426 -13.47016319]


- b0= 158.4913 
- b1= -1.1416
- b2= -0.4420
- b3= -13.47016319
### Y= 158.4913 -1.1416X1-0.4420X2-13.47016319X3

In [25]:
Y_hat= 158.4913 -1.1416*X1-0.4420*X2-13.47016319*X3

In [34]:
allX = [X1, X2, X3]

In [37]:
for x in allX:
    x_bar = np.mean(x)
    Y_bar = np.mean(Y)
    x_err = x - x_bar
    Y_err = Y - Y_bar
    A = np.sum(x_err*Y_err)
    B = np.sum(x_err**2)
    b1 = A / B
    b0 = Y_bar - b1*x_bar
    n = len(x)
    Y_hat = b0 + b1*x
    resid = Y - Y_hat
    SSE = np.sum((Y - Y_hat)**2)
    MSE = SSE / (n-2)
    SSR = sum((Y_hat - Y_bar)**2)
    SSTO = SSR + SSE
    R_squared = SSR/SSTO
    print(f' R_square {R_squared}')
    print(f' MSE {MSE}')
    print(f' SSE {SSE}')
    print(f' SSTO {SSTO}')
    print(f' SSR {SSR}')

 R_square 0.6189842519960209
 MSE 115.77080673590463
 SSE 5093.915496379804
 SSTO 13369.304347826084
 SSR 8275.38885144628
 R_square 0.3635387359110574
 MSE 193.38737148198152
 SSE 8509.044345207187
 SSTO 13369.304347826084
 SSR 4860.260002618898
 R_square 0.41549754587804455
 MSE 177.59980002744717
 SSE 7814.391201207676
 SSTO 13369.304347826086
 SSR 5554.91314661841


In [39]:
import statsmodels.api as sm
import statsmodels.formula.api as smf
model0 = smf.ols('Y~ X1+X2',data=df).fit()
table0 = sm.stats.anova_lm(model0, typ=3)
print(table0)

                sum_sq    df          F        PR(>F)
Intercept  7579.172193   1.0  70.649120  1.273843e-10
X1         3896.044142   1.0  36.316907  3.347580e-07
X2          480.915294   1.0   4.482843  4.005967e-02
Residual   4613.000203  43.0        NaN           NaN


In [41]:
import statsmodels.api as sm
import statsmodels.formula.api as smf
model1 = smf.ols('Y~ X1+X2+X3',data=df).fit()
table1 = sm.stats.anova_lm(model1, typ=3)
print(table1)

                sum_sq    df          F        PR(>F)
Intercept  7734.515725   1.0  76.456070  5.260955e-11
X1         2857.553377   1.0  28.247056  3.810252e-06
X2           81.659053   1.0   0.807204  3.740702e-01
X3          364.159521   1.0   3.599735  6.467813e-02
Residual   4248.840682  42.0        NaN           NaN


In [42]:
F_star123 = ((4613.000203-4248.840682)/(43-42))/(4248.840682  /42)
F_star123

3.5997348516255765

In [44]:
F123 = stats.f.ppf(q = 1-0.02, dfn=2, dfd=len(X1)-4)
F123

4.300122750135479

### Before jumping straight into the conclusion, I would like to state the decision rule.
- H0 : β4 = 0
- H1 : β4 != 0
### Since the F-star= 3.59< F= 4.3 , we can reject H0 and conclude Ha, and conclude that the presence of X3 is not significant in the data set that already included X1 and X2 and can be excluded from the data

### b

### Subset with two predictors

In [45]:
X12 = df.iloc[:,1:3]
X12 = sm.add_constant(X12)
X13 = df.iloc[:,[1,3]]
X13 = sm.add_constant(X13)
X23 = df.iloc[:,2:4]
X23 = sm.add_constant(X23)
X_double = [X12, X23, X13]

In [48]:
for x in allX:
    x_bar = np.mean(x)
    Y_bar = np.mean(Y)
    x_err = x - x_bar
    Y_err = Y - Y_bar
    A = np.sum(x_err*Y_err)
    B = np.sum(x_err**2)
    b1 = A / B
    b0 = Y_bar - b1*x_bar
    n = len(x)
    Y_hat = b0 + b1*x
    resid = Y - Y_hat
    SSE = np.sum((Y - Y_hat)**2)
    MSE = SSE / (n-2)
    SSR = sum((Y_hat - Y_bar)**2)
    SSTO = SSR + SSE
    R_squared = SSR/SSTO
    R_2a = 1 - (((n-1)/(n-3))*(SSE/SSTO))
    print(f' Adj R_square {R_2a}')
    

 Adj R_square 0.6012625892981616
 Adj R_square 0.3339358864185483
 Adj R_square 0.38831138522120934


### Conclusion
- With Ra_square : Model X12 has highest number = 0.601. So, The best subset of predictorvariable which is recommended is the one that contain X1 and X2

### C

In [50]:
X = df.iloc[:,1:4]
X = sm.add_constant(X)

In [51]:
X_trans = np.transpose(X)
X_transX = X_trans.dot(X)
X_transY = X_trans.dot(Y)
Invert = np.linalg.pinv(X_transX)
bb = Invert.dot(X_transY)
Y_hat = X.dot(bb)
J = np.ones((len(X1), len(X1)))
b_trans = np.transpose(bb)
resid = Y - Y_hat
Y_trans = np.transpose(Y)
Y_transY = Y_trans.dot(Y)
Y_transJ = Y_trans.dot(J)
YJY = Y_transJ.dot(Y)
b_transXY = b_trans.dot(X_transY)
SSTO = Y_transY - ((1/len(X1))*YJY)
SSE = Y_transY - b_transXY
MSE = SSE / (len(X1)-4)
print(bb)

[158.49125167  -1.14161185  -0.44200426 -13.47016319]


In [54]:

from statsmodels.stats.outliers_influence import OLSInfluence
model = sm.OLS(Y, X, data = df)
results = model.fit()
results.summary()
test_class = OLSInfluence(results)
dir(test_class)
h = test_class.hat_matrix_diag
aa = []

C:\Users\tanvu\anaconda3\lib\site-packages\statsmodels\base\model.py:127: ValueWarning: unknown kwargs ['data']
  warnings.warn(msg, ValueWarning)


In [55]:
ti = resid * (((len(X1) - 4 - 1)/(SSE*(1-h)-(resid**2)))**(1/2))
ti

0     0.011555
1    -0.933179
2     0.403625
3     0.214666
4     0.594417
5    -0.381671
6    -1.306938
7     0.043505
8    -0.729007
9     0.202832
10    1.835846
11    1.362219
12   -1.498367
13   -1.558052
14    1.357623
15   -0.281510
16    1.806742
17    0.874528
18   -1.102377
19    0.867553
20    0.573154
21    0.901148
22    0.368162
23   -0.409310
24   -0.476760
25   -0.432335
26   -1.974202
27    0.589891
28   -0.988725
29    0.347340
30   -1.769416
31    1.219938
32    0.061648
33   -1.542156
34    0.095941
35    1.176307
36    1.227789
37   -0.549421
38   -0.986985
39   -0.589772
40    1.118953
41   -0.095402
42   -1.422228
43    1.345415
44   -0.567127
45    1.044921
dtype: float64

In [69]:
# Bonferroni outlier test procedure: t(1 − α/2n; n − p − 1)
n = len(X1)
p = 4
alpha = 0.05
t = stats.t.ppf(1 - alpha/(2*n), n - p - 1)
t


3.5153486707360724

In [71]:
for i in range(46):
 if ti[i] > t:
    print(b[i])
else:
 print('|ti| always lower than t')

|ti| always lower than t


### THERE IS OUTLIER IN THIS CASE AS Ti always lower than t

### D

In [59]:
for i in h:
 aa.append(i)
print(aa)

[0.07819669444566327, 0.06706793454478491, 0.03717097449286008, 0.1536108379033404, 0.09673692469196835, 0.12857668418008095, 0.03448499938222793, 0.07524431346919982, 0.18425851338615706, 0.05797909857306274, 0.08759237323467639, 0.030874655190255344, 0.09032064425107117, 0.03323759569283374, 0.1428903222948928, 0.04713297229132751, 0.11954225591636203, 0.0624173773647996, 0.03350766536337926, 0.12892850579463183, 0.07769552730248716, 0.13690056195089323, 0.03288050497795675, 0.13575135256026258, 0.043367319937463664, 0.10294630245201468, 0.08682305129329086, 0.18601918951143726, 0.05944210499883176, 0.08998055581822445, 0.11710546466583385, 0.10963099113639307, 0.04504471384292162, 0.03717097449286008, 0.10303976954311114, 0.027232297500199502, 0.121220910858725, 0.07058922949004309, 0.1809601031131856, 0.086895984622751, 0.03797571645008457, 0.15385909185944552, 0.06101915346885889, 0.050909576788637334, 0.07261644156528402, 0.08315176733522914]


In [72]:
mean_hatval = 4/46
cutoff = 2 * mean_hatval
print(mean_hatval, cutoff)

0.08695652173913043 0.17391304347826086


In [73]:
for i in range(46):
    if aa[i] > cutoff:
       print(b[i])


9
28
39


### E

In [75]:
# Cook's distance
Di = ((resid**2)/(4*MSE))*(h/((1-h)**2))
Di.head(5)

0    0.000003
1    0.015699
2    0.001604
3    0.002139
4    0.009608
dtype: float64

In [76]:
for i in range(46):
    if abs(Di[i]) > mean_hatval:
       print(b[i])

17
31


In [78]:
dffits = ti * ((h / (1 - h))**(1/2))

In [80]:
cutoff1 = 2 * np.sqrt(mean_hatval)
for i in range(46):
    if abs(dffits[i]) > cutoff1:
       print(b[i])

17
27
31
